src/Npgsql/MetricsReporter.cs

```diff
@@ -0,0 +1,238 @@
+        => ConnectionTimeouts.Add(1, _poolNameTag);
+
+    internal void ReportPendingConnectionRequestStart()
+        => PendingConnectionRequests.Add(1, _poolNameTag);
+    internal void ReportPendingConnectionRequestStop()
+        => PendingConnectionRequests.Add(-1, _poolNameTag);
+
+    internal void ReportConnectionCreateTime(TimeSpan duration)
+        => ConnectionCreateTime.Record(duration.TotalMilliseconds, _poolNameTag);
+
+    static IEnumerable<Measurement<int>> GetConnectionUsage()
+    {
+        lock (Reporters)
+        {
+            var measurements = new List<Measurement<int>>();
+
+            for (var i = 0; i < Reporters.Count; i++)
+            {
+                var reporter = Reporters[i];
+
+                if (reporter._dataSource is PoolingDataSource poolingDataSource)
+                {
+                    var stats = poolingDataSource.Statistics;
+
+                    measurements.Add(new Measurement<int>(
+                        stats.Idle,
+                        reporter._poolNameTag,
+                        new KeyValuePair<string, object?>("state", "idle")));
+
+                    measurements.Add(new Measurement<int>(
+                        stats.Busy,
+                        reporter._poolNameTag,
+                        new KeyValuePair<string, object?>("state", "used")));
+                }
+            }
+
+            return measurements;
+        }
+    }
+
+    static IEnumerable<Measurement<int>> GetMaxConnections()
+    {
+        lock (Reporters)
+        {
+            var measurements = new List<Measurement<int>>();
+
+            foreach (var reporter in Reporters)
+            {
+                if (reporter._dataSource is PoolingDataSource poolingDataSource)
+                {
+                    measurements.Add(new Measurement<int>(poolingDataSource.MaxConnections, reporter._poolNameTag));
+                }
+            }
+
+            return measurements;
+        }
+    }
+
+    static IEnumerable<Measurement<double>> GetPreparedCommandsRatio()
+    {
+        lock (Reporters)
+        {
+            var measurements = new Measurement<double>[Reporters.Count];
+
+            for (var i = 0; i < Reporters.Count; i++)
+            {
+                var reporter = Reporters[i];
+
+                var counters = new CommandCounters
+                {
+                    All = Interlocked.Exchange(ref reporter._commandCounters.All, default)
+                };
+
+                measurements[i] = new Measurement<double>(
+                    (double)counters.PreparedCommandsStarted / counters.CommandsStarted * 100,
+                    reporter._poolNameTag);
+            }
+
+            return measurements;
+        }
+    }
+
+    public void Dispose()
+    {
+        lock (Reporters)
+        {
+            Reporters.Remove(this);
+            Reporters.Sort((x,y) => string.Compare(x._dataSource.Name, y._dataSource.Name, StringComparison.Ordinal));
```

Comment:
Don't really need to sort since it's already sorted.

Reply:
How so? The idea here is to provide deterministic alphabetical ordering, but Reporters is just a List to which we add based on which data source got instantiated first, no?

(in any case, add/remove should be a really rare operation so we shouldn't care about perf)

Reply:
Ah sure, just for the removing, OK.